src/ray/gcs/pubsub/gcs_pub_sub.h

```diff
@@ -0,0 +1,111 @@
+// Copyright 2017 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef RAY_GCS_GCS_PUB_SUB_H_
+#define RAY_GCS_GCS_PUB_SUB_H_
+
+#include "ray/common/id.h"
+#include "ray/gcs/callback.h"
+#include "ray/gcs/redis_client.h"
+#include "ray/protobuf/gcs.pb.h"
+
+namespace ray {
+namespace gcs {
+
+using rpc::ActorTableData;
+using rpc::GcsNodeInfo;
+using rpc::HeartbeatBatchTableData;
+using rpc::HeartbeatTableData;
+using rpc::JobTableData;
+using rpc::ObjectChange;
+using rpc::ResourceChange;
+using rpc::TablePubsub;
+using rpc::TaskLeaseData;
+using rpc::TaskTableData;
+using rpc::WorkerFailureData;
+
+/// \class GcsPubSub
+///
+/// GcsPubSub supports publishing, subscription and unsubscribing of data.
+template <typename ID, typename Data>
+class GcsPubSub {
+ public:
+  /// The callback is called when a subscription message is received.
+  using Callback = std::function<void(const ID &id, const Data &data)>;
+
+  explicit GcsPubSub(std::shared_ptr<RedisClient> redis_client)
+      : redis_client_(redis_client) {}
+
+  virtual ~GcsPubSub() = default;
+
+  /// Posts a message to the given channel.
+  ///
+  /// \param channel The channel to publish to redis.
+  /// \param id The id of message to be published to redis.
+  /// \param data The data of message to be published to redis.
+  /// \param done Callback that will be called when the message is published to redis.
+  /// \return Status
+  Status Publish(const std::string &channel, const ID &id, const Data &data,
```

Comment:
I'd be better to define a enum for the channel.

Reply:
fixed